### ISSUE TITLE
Failed to build for iOS 8.4 using XCode 6.4

### ISSUE THREAD

**Issue by Gowzancha** [randombit/botan#188]
I failed to build Botan 1.11.17 for iOS 8.4. According to the tutorial I executed following commands:
1. $ ./configure.py --cpu=armv7 --cc=clang --cc-abi-flags="-arch armv7 -arch armv7s --sysroot=/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS8.4.sdk"

2. I edit makefile:
AR = libtool -static -o

3. $ make
clang: error: no such file or directory: 'armv7'
clang: error: no such file or directory: 'armv7s'
clang: error: invalid arch name '-arch -pthread'
make: *** [build/obj/lib/asn1_alg_id.o] Error 1

I got similar errors when I try to build it for simulator (i386) too. any idea?

**Comment by webmaster128** [comment]
It would be interesting to see which clang command is run by make before that error. Could you add more output please? Maybe linked as a Gist.

**Comment by Gowzancha** [comment]
https://gist.github.com/Gowzancha/14a4a8453a3461e052f7

*since I'm new here I'm not sure it's the clean way to link a gist, sorry)

**Comment by webmaster128** [comment]
You are highly welcome! Gist is just an integrated pastebin in Github.

The line I was looking for is

```
clang++ --sysroot=/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS8.4.sdk -arch -pthread armv7 armv7s -fPIC -fvisibility=hidden -std=c++11 -D_REENTRANT -fstack-protector -O3 -Wall -Wextra -Wstrict-aliasing -Wstrict-overflow=5 -Wcast-align -Wmissing-declarations -Wpointer-arith -Wcast-qual -Wunreachable-code -Ibuild/include -c ./src/lib/asn1/alg_id.cpp -o build/obj/lib/asn1_alg_id.o
```

As you can see after `-arch`, your arguments are somehow merged and the the process `-pthread` is put in betweeen.

I don't know how that happened but you could manually edit your Makefile (after the `./configure.py` run) in order to clean up the arguments.

Ps.: If you are starting a new application, I highly recommend using the latest Botan version 1.11.18.

**Comment by cordney** [comment]
I can reproduce this behaviour with 1.11.18 and master. The Makefile generator somehow seems to mess up with the `--cc-abi-flags` option. 

```
./configure.py --cpu=armv7 --cc=clang --cc-abi-flags="-arch armv7 -arch armv7s --sysroot=/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS8.4.sdk/"
```

generates the Makefile:

```
# Compiler Options
CXX            = clang++ --sysroot=/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS8.4.sdk/ -arch -pthread armv7 armv7s
LINKER         = $(CXX)
LIB_OPT        = -c -O3
APP_OPT        = -c -O2
LANG_FLAGS     = -std=c++11 -D_REENTRANT -fstack-protector
WARN_FLAGS     = -Wall -Wextra -Wstrict-aliasing -Wstrict-overflow=5 -Wcast-align -Wmissing-declarations -Wpointer-arith -Wcast-qual -Wunreachable-code
SO_OBJ_FLAGS   = -fPIC -fvisibility=hidden

(remainder ommitted)
```

Instead of `-arch armv7 -arch armv7s` it generates `-arch -pthread armv7 armv7s`. I don't know where the `-pthread` comes from, but it seems unnecessary in the CXX flags here, too.

After manually fixing this, I get errors about missing C+11 headers, such as `<cstdint>`. According to https://github.com/genome/joinx/issues/3 and other sources, you have to use libc++ instead of libstdc++, so adding `-stdlib=libc++`to CXX flags solves the problem.

Given Travis CI supports iOS 8.4 SDK (see http://blog.travis-ci.com/2015-07-08-xcode-6-4-now-available/), I think we could easily add and iOS build to the CI scripts for the future to avoid such regressions (after we fixed the issues here).

**Comment by cordney** [comment]
I could track down the offending lines to https://github.com/randombit/botan/blob/master/configure.py#L851 and https://github.com/randombit/botan/blob/master/configure.py#L863. First, a set is used to store the ABI link flags, which does not allow multiple occurrences of the same flag, e.g., `-arch armv7 -arch armv7`. Second, the set is converted to a list and then sorted alphabetically, which puts `-pthread` in between `-arch` and `armv7`. I fixed both issues in 69e2654, which @randombit already merged.
